base/inc/AdePT/VolumeDispatcher.h

```diff
@@ -0,0 +1,86 @@
+// SPDX-FileCopyrightText: 2021 CERN
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * @file VolumeDispatcher.h
+ * @brief Dispatching queries to volumes without virtual calls.
+ */
+
+#ifndef VOLUME_DISPATCHER_H_
+#define VOLUME_DISPATCHER_H_
+
+#include <CopCore/Global.h>
+
+#include <VecGeom/base/Vector3D.h>
+#include <VecGeom/volumes/PlacedVolume.h>
+#include <VecGeom/volumes/VolumeTypes.h>
+
+#include <VecGeom/volumes/Box.h>
+#include <VecGeom/volumes/Trd.h>
+#include <VecGeom/volumes/Tube.h>
+
+#define DISPATCH_TO_PLACED_VOLUME_KIND(vol, kind, class, impl, type, func, ...)                \
+  case vecgeom::VolumeTypes::kind: {                                                           \
+    type output;                                                                               \
+    const auto unplacedStruct = static_cast<const vecgeom::class *>(vol)->GetUnplacedStruct(); \
+    vecgeom::impl::func(*unplacedStruct, __VA_ARGS__, output);                                 \
+    return output;                                                                             \
+  }
+
+#define DISPATCH_TO_PLACED_VOLUME(vol, type, func, ...)                                                            \
+  switch (vol->GetType()) {                                                                                        \
+    DISPATCH_TO_PLACED_VOLUME_KIND(vol, kBox, PlacedBox, BoxImplementation, type, func, __VA_ARGS__);              \
+    DISPATCH_TO_PLACED_VOLUME_KIND(vol, kTrd, PlacedTrd, TrdImplementation<vecgeom::TrdTypes::UniversalTrd>, type, \
```

Comment:
is there an additional benefit, in comparison to (let's say for the box case): 
```
static_cast<PlacedBox*>(vol)->PlacedBox::FUNC
```

In such version, one would not need to know/add the "Implementation" type since it's known internally.

Reply:
Calling `PlacedBox::FUNC` doesn't work because `PlacedBox` is an abstract class and doesn't provide implementations for `Contains`, `DistanceToIn`, etc. - that only comes with `SpecializedBox` (and is the reason why I brought up the complex inheritance structure in the last meeting). Qualifying to `SimpleBox` works (see #101), but I think this leaks even more internals about VecGeom.